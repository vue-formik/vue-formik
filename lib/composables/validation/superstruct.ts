import { type Struct, assert, StructError } from "superstruct";
import { setNestedValue } from "@/helpers";
import { constructPath } from "@/helpers/formikObject/objectPath";

const validateSuperstruct = <T extends object>(values: T, schema: Struct<T>) => {
  const errors = {};
  try {
    assert(values, schema);
  } catch (e) {
    if (e instanceof StructError) {
      const failures = e.failures();
      console.log(failures, "---");
      for (const failure of failures) {
        const { path, message } = failure;
        setNestedValue(errors, constructPath(path), message);
      }
    }
  }
  return errors;
};

export default validateSuperstruct;
