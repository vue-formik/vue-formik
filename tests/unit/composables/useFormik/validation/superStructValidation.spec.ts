import { describe, test, expect } from "vitest";
import * as s from "superstruct";
import { useFormik } from "@/index";
import { nextTick } from "vue";

describe("superstruct validation", () => {
  const userSchema = s.object({
    name: s.nonempty(s.string()),
    age: s.optional(s.number()),
    email: s.pattern(s.string(), /^[^\s@]+@[^\s@]+\.[^\s@]+$/),
    hobbies: s.optional(s.array(s.nonempty(s.string()))),
    address: s.optional(
      s.object({
        street: s.optional(s.size(s.string(), 2, 32)),
        city: s.optional(s.size(s.string(), 2, 16)),
      }),
    ),
  });

  const user = {
    name: "John Doe",
    age: 30,
    email: "test",
    hobbies: ["reading", "gaming"],
    address: {},
  } as s.Infer<typeof userSchema>;

  test("validate string", () => {
    const fk = useFormik({
      initialValues: user,
      structSchema: userSchema,
      validateOnMount: true,
    });

    expect(fk.isValid.value).toBe(false);
    expect(fk.errors).toMatchSnapshot();
  });

  test("validate array", async () => {
    const fk = useFormik({
      initialValues: {
        ...user,
        email: "test@test.com",
      },
      structSchema: userSchema,
    });
    fk.setFieldValue("hobbies[0]", "");
    fk.setFieldTouched("hobbies[0]", true);

    await nextTick();

    expect(fk.isValid.value).toBe(false);
    expect(fk.errors).toMatchSnapshot();
  });

  test("validate object", async () => {
    const fk = useFormik({
      initialValues: user,
      structSchema: userSchema,
    });

    fk.setFieldValue("address.street", "A");
    fk.setFieldTouched("address.street", true);

    await nextTick();

    expect(fk.isValid.value).toBe(false);
    expect(fk.errors).toMatchSnapshot();
  });

  test("complex validation", async () => {
    const fk = useFormik({
      initialValues: {
        criteria: [{ property: "", comparator: "", range: [{ from: "", to: "" }] }],
      },
      structSchema: s.object({
        criteria: s.array(
          s.object({
            property: s.optional(s.string()),
            comparator: s.optional(s.string()),
            range: s.optional(
              s.array(
                s.object({
                  from: s.nonempty(s.string()),
                  to: s.nonempty(s.string()),
                }),
              ),
            ),
          }),
        ),
      }),
    });

    fk.setFieldValue("criteria[0].property", "test");

    await nextTick();

    expect(fk.isValid.value).toBe(false);
    expect(fk.errors).toMatchSnapshot();
  });
});
