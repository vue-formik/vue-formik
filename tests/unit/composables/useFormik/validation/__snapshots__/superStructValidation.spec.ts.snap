// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`superstruct validation > complex validation 1`] = `
{
  "criteria": [
    {
      "range": [
        {
          "from": "Expected a nonempty string but received an empty one",
          "to": "Expected a nonempty string but received an empty one",
        },
      ],
    },
  ],
}
`;

exports[`superstruct validation > validate array 1`] = `
{
  "hobbies": [
    "Expected a nonempty string but received an empty one",
  ],
}
`;

exports[`superstruct validation > validate object 1`] = `
{
  "address": {
    "street": "Expected a string with a length between \`2\` and \`32\` but received one with a length of \`1\`",
  },
  "email": "Expected a string matching \`/^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$/\` but received "test"",
}
`;

exports[`superstruct validation > validate string 1`] = `
{
  "email": "Expected a string matching \`/^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$/\` but received "test"",
}
`;
